### PR TITLE
implementing keys for JuMPArray

### LINKED ISSUE #1579: keys(::JuMPArray) bugs with 1 dimension
On master, I get an error on keys with a JuMPArray of 1 dimension, but I don't get the error with 2 or more dimensions.

```julia
julia> begin
           m = Model()
           @variable(m, 0 <= x[2:3] <= 1)
           @show typeof(x)
           @show keys(x)
       end
````
````
typeof(x) = JuMPArray{VariableRef,1,Tuple{UnitRange{Int64}},Tuple{Dict{Int64,Int64}}}
ERROR: JuMPArray does not support this operation.
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] LinearIndices(::JuMPArray{VariableRef,1,Tuple{UnitRange{Int64}},Tuple{Dict{Int64,Int64}}}) at /Users/itahiri/.julia/packages/JuMP/uAycC/src/JuMPArray.jl:96
 [3] keys(::JuMPArray{VariableRef,1,Tuple{UnitRange{Int64}},Tuple{Dict{Int64,Int64}}}) at ./abstractarray.jl:100
 [4] macro expansion at /Users/itahiri/.julia/packages/JuMP/uAycC/src/macros.jl:301 [inlined]
 [5] top-level scope at ./REPL[26]:3
````

```julia
julia> begin
           m = Model()
           @variable(m, 0 <= x[2:3, 2:3] <= 1)
           @show typeof(x)
           @show keys(x)
       end
````
````
typeof(x) = JuMPArray{VariableRef,2,Tuple{UnitRange{Int64},UnitRange{Int64}},Tuple{Dict{Int64,Int64},Dict{Int64,Int64}}}
keys(x) = CartesianIndex{2}[CartesianIndex(2, 2) CartesianIndex(2, 3); CartesianIndex(3, 2) CartesianIndex(3, 3)]
2×2 CartesianIndices{2,Tuple{UnitRange{Int64},UnitRange{Int64}}}:
 CartesianIndex(2, 2)  CartesianIndex(2, 3)
 CartesianIndex(3, 2)  CartesianIndex(3, 3)
````

### PR DESCRIPTION
**Pull request by IssamT**
closes #1579

### CODE REVIEW

File: src/JuMPArray.jl [hunk 1]
```diff
@@ -97,6 +97,20 @@ else
     Base.size(A::JuMPArray) = size(A.data)
     Base.LinearIndices(A::JuMPArray) = error("JuMPArray does not support this operation.")
     Base.axes(A::JuMPArray) = A.axes
+    Base.CartesianIndices(a::JuMPArray) = CartesianIndices(a.data)
+    struct JuMPArrayKey{T<:Tuple}
```
Comment (mlubin) [inline src/JuMPArray.jl]: Please add a docstring.

File: test/containers.jl [hunk 2]
```diff
@@ -247,4 +247,15 @@ And data, a 2×2×2×2 Array{Float64,4}:
 And data, a 0-dimensional Array{$Int,0}:
 1"""
     end
+
+    @testset "JuMPArray keys" begin
```
Comment (mlubin) [inline test/containers.jl]: Do these tests pass on 0.6?

File: src/JuMPArray.jl [hunk 3]
```diff
@@ -97,6 +97,20 @@ else
...
+        I::T
+    end
+    Base.getindex(k::JuMPArrayKey, index::Int) = getindex(k.I, index)
+    function JuMPArrayKey(ci::CartesianIndex, axes_array)
+        return JuMPArrayKey(Tuple(map(x->axes_array[x[1]][x[2]],
+                            enumerate(ci.I))))
+    end
+    function Base.keys(a::JuMPArray)
+        axes_array = map(x->collect(x), a.axes)
```
Comment (mlubin) [inline src/JuMPArray.jl]: Do we need to collect the axes? That's potentially very expensive.

Comment (mlubin) [inline src/JuMPArray.jl]: Nit: `x -> collect(x)` can be replaced by `collect`

File: src/JuMPArray.jl [hunk 4]
```diff
@@ -97,6 +97,20 @@ else
     Base.size(A::JuMPArray) = size(A.data)
     Base.LinearIndices(A::JuMPArray) = error("JuMPArray does not support this operation.")
     Base.axes(A::JuMPArray) = A.axes
+    Base.CartesianIndices(a::JuMPArray) = CartesianIndices(a.data)
+    struct JuMPArrayKey{T<:Tuple}
+        I::T
+    end
+    Base.getindex(k::JuMPArrayKey, index::Int) = getindex(k.I, index)
+    function JuMPArrayKey(ci::CartesianIndex, axes_array)
```
Comment (mlubin) [inline src/JuMPArray.jl]: I don't follow this function very easily. Could you add a comment explaining what it does?

File: src/JuMPArray.jl [hunk 5]
```diff
@@ -97,6 +97,30 @@ else
...
+        # using axes_arrays to translate elements of an index of
+        # JuMPArray data into elements of its axes, in order
+        # to create the JuMPArrayKey.
+        return JuMPArrayKey(Tuple(map(x->axes_array[x[1]][x[2]],
+                            enumerate(ci.I))))
+    end
+    as_array(collection) = collect(collection)
+    as_array(collection::AbstractArray) = collection
+    function Base.keys(a::JuMPArray)
```
[the remark this answers is not in the document]
Comment (IssamT) [reply src/JuMPArray.jl]: You're right. It's fixed.

File: src/JuMPArray.jl [hunk 6]
```diff
@@ -116,9 +116,27 @@ else
...
+        next = iterate(iter.cartesian_indices, state)
+        return next == nothing ? nothing :
+               (JuMPArrayKey(next[1], iter.axes_array), next[2])
+    end
+    Base.length(iter::JuMPArrayKeys) = length(iter.cartesian_indices)
+    function Base.eltype(iter::JuMPArrayKeys)
+        return JuMPArrayKey{Tuple{eltype.(iter.axes_array)...}}
+    end
     function Base.keys(a::JuMPArray)
         axes_array = map(as_array, a.axes)
```
Comment (mlubin) [inline src/JuMPArray.jl]: If you use `Iterators.product` over `a.axes`, could you avoid this `as_array` step? There would be no need to touch `CartesianIndices` either.

**Pull request merged**